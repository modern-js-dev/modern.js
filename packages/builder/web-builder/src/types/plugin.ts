import type { WebBuilderConfig } from './config';
import type { WebBuilderContext } from './context';
import type {
  OnExitFn,
  OnAfterBuildFn,
  OnBeforeBuildFn,
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyBuilderConfigFn,
  OnAfterCreateCompilerFn,
  OnBeforeCreateCompilerFn,
} from './hooks';

export type PluginStore = {
  plugins: WebBuilderPlugin[];
  addPlugins: (plugins: WebBuilderPlugin[]) => void;
  removePlugins: (pluginNames: string[]) => void;
  isPluginExists: (pluginName: string) => boolean;
};

export type WebBuilderPluginAPI = {
  context: WebBuilderContext;
  isPluginExists: PluginStore['isPluginExists'];
  getBuilderConfig: () => WebBuilderConfig;

  // Hooks
  onExit: (fn: OnExitFn) => void;
  onAfterBuild: (fn: OnAfterBuildFn) => void;
  onBeforeBuild: (fn: OnBeforeBuildFn) => void;
  onAfterCreateCompiler: (fn: OnAfterCreateCompilerFn) => void;
  onBeforeCreateCompiler: (fn: OnBeforeCreateCompilerFn) => void;

  // Modifiers
  modifyWebpackChain: (fn: ModifyWebpackChainFn) => void;
  modifyWebpackConfig: (fn: ModifyWebpackConfigFn) => void;
  modifyBuilderConfig: (fn: ModifyBuilderConfigFn) => void;
};

export type WebBuilderPlugin = {
  name: string;
  setup: (api: WebBuilderPluginAPI) => Promise<void> | void;
};
