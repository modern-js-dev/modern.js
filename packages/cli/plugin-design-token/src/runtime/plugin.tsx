import type { Plugin } from '@modern-js/runtime-core';
import React, { useContext } from 'react';

export const DesignTokenContext = React.createContext<any>({});

export const useDesignTokens = (): IDesignTokens =>
  useContext<IDesignTokens>(DesignTokenContext);

export default (
  options: {
    token?: Record<string, any>;
    useStyledComponentsThemeProvider?: boolean;
    useDesignTokenContext?: boolean;
  } = {},
): Plugin => ({
  name: '@modern-js/plugin-design-token',

  setup: () => ({
    hoc({ App }, next) {
      return next({
        App: (props: any) => {
          const {
            token = {},
            useStyledComponentsThemeProvider = false,
            useDesignTokenContext = false,
          } = options;
          if (useStyledComponentsThemeProvider && useDesignTokenContext) {
            const { ThemeProvider } = require('@modern-js/runtime-core/styled');
            ThemeProvider.init =
              (App as React.ComponentType<any> & { init: () => void }).init ||
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              function () {};
            return (
              <ThemeProvider theme={token}>
                <DesignTokenContext.Provider value={token}>
                  <App {...props} />
                </DesignTokenContext.Provider>
              </ThemeProvider>
            );
          } else if (useStyledComponentsThemeProvider) {
            const { ThemeProvider } = require('@modern-js/runtime-core/styled');
            ThemeProvider.init =
              (App as React.ComponentType<any> & { init: () => void }).init ||
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              function () {};
            return (
              <ThemeProvider theme={token}>
                <App {...props} />
              </ThemeProvider>
            );
          } else if (useDesignTokenContext) {
            (
              DesignTokenContext as React.Context<any> & { init: () => void }
            ).init =
              (App as React.ComponentType<any> & { init: () => void }).init ||
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              function () {};
            return (
              <DesignTokenContext.Provider value={token}>
                <App {...props} />
              </DesignTokenContext.Provider>
            );
          } else {
            return <App {...props} />;
          }
        },
      });
    },
  }),
});
