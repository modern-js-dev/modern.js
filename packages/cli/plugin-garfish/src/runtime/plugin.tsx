import { createPlugin } from '@modern-js/runtime-core';
import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { logger } from '../util';
import { GarfishProvider } from './utils/Context';
import setExternal from './utils/setExternal';
import {
  Config,
  Manifest,
  MicroComponentProps,
  ModulesInfo,
  Options,
} from './useModuleApps';
import { generateMApp } from './utils/MApp';
import { AppMap, generateApps } from './utils/apps';

async function initOptions(manifest: Manifest = {}, options: Options) {
  let apps: ModulesInfo = [];

  // use manifest modules
  if (manifest?.modules) {
    apps = manifest?.modules;
    logger('manifest modules', apps);
  }

  // get module list
  if (manifest?.getAppList) {
    apps = await manifest?.getAppList();
    logger('getAppList modules', apps);
  }

  // get inject modules list
  if (window?.modern_manifest?.modules) {
    apps = window?.modern_manifest?.modules;
    logger('modern_manifest', apps);
  }

  return {
    apps,
    ...options,
  };
}

export default ((config: Config) => {
  setExternal();
  const { manifest = {}, LoadingComponent, ...options } = config;
  logger('createPlugin', { config });
  if (!manifest.LoadingComponent && LoadingComponent) {
    manifest.LoadingComponent = LoadingComponent;
  }

  const promise = initOptions(manifest, options);

  return createPlugin(() => ({
    hoc({ App }, next) {
      class GetMicroFrontendApp extends React.Component {
        state: {
          MApp: React.FC<MicroComponentProps>;
          apps: AppMap;
          appInfoList: ModulesInfo;
        } = {
          MApp: () => {
            logger('MApp init Component Render');
            return React.createElement('div');
          },
          apps: new Proxy(
            {},
            {
              get(_target, _p) {
                logger('apps init Component Render', _p);
                return () => React.createElement('div');
              },
            },
          ),
          appInfoList: [],
        };

        constructor(props: any) {
          super(props);
          const load = async () => {
            const GarfishConfig = await promise;
            logger('initOptions result', { manifest, GarfishConfig });
            const MApp = generateMApp(GarfishConfig, manifest);
            const { appInfoList, apps } = generateApps(GarfishConfig, manifest);
            logger('generateApps', { MApp, apps, appInfoList });
            this.setState({
              MApp,
              apps,
              appInfoList,
            });
          };
          load();
        }

        render() {
          logger('GarfishProvider state', this.state);
          return (
            <GarfishProvider value={this.state}>
              <App {...this.props} {...this.state} />
            </GarfishProvider>
          );
        }
      }

      return next({
        App: hoistNonReactStatics(GetMicroFrontendApp, App),
      });
    },
  }));
}) as any;
