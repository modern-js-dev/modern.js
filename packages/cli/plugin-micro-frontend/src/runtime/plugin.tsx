import { createPlugin } from '@modern-js/runtime-core';
import { useMemo } from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { GarfishProvider } from './utils/Context';
import Apps from './utils/apps';
import setExternal from './utils/setExternal';
import { Config } from './typings';

export default ((config: Config) => {
  setExternal();
  const appsInstance = new Apps(config);

  return createPlugin(() => ({
    hoc({ App }, next) {
      appsInstance.run();

      const getMicroFrontendApp = (props: any) => {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const value = useMemo(
          () => ({
            MApp: appsInstance.getMApp(),
            apps: appsInstance.getApps(),
          }),
          [],
        );

        return (
          <GarfishProvider value={value}>
            <App {...props} />
          </GarfishProvider>
        );
      };
      return next({
        App: hoistNonReactStatics(getMicroFrontendApp, App),
      });
    },
  }));
}) as any;
