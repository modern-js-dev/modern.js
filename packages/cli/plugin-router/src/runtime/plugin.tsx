import React, { useContext } from 'react';
import {
  createBrowserHistory,
  createHashHistory,
  History,
  BrowserHistoryBuildOptions,
  HashHistoryBuildOptions,
} from 'history';
import { Router, StaticRouter, RouteProps } from 'react-router-dom';
import { createPlugin, RuntimeReactContext } from '@modern-js/runtime-core';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { resolveBasename, renderRoutes, getLocation } from './utils';

export type SingleRouteConfig = RouteProps & {
  redirect?: string;
  routes?: SingleRouteConfig[];
  key?: string;

  /**
   * layout component
   */
  layout?: React.ComponentType;

  /**
   * component would be rendered when route macthed.
   */
  component?: React.ComponentType;
};

export type HistoryConfig =
  | {
      supportHtml5History: true;
      historyOptions: BrowserHistoryBuildOptions;
    }
  | {
      supportHtml5History: false;
      historyOptions: HashHistoryBuildOptions;
    };

export type RouterConfig = Partial<HistoryConfig> & {
  routesConfig?: {
    globalApp?: React.ComponentType<any>;
    routes?: SingleRouteConfig[];
  };
  history?: History;
};

// todo: check
const isBrowser = () => typeof window !== 'undefined';

export const routerPlugin: any = ({
  history: customHistory,
  supportHtml5History = true,
  routesConfig,
  historyOptions,
}: RouterConfig) => {
  const isBrow = isBrowser();
  // eslint-disable-next-line no-nested-ternary
  const history = isBrow
    ? customHistory || supportHtml5History
      ? createBrowserHistory(historyOptions)
      : createHashHistory(historyOptions)
    : ({} as History);

  return createPlugin(
    () => ({
      hoc: ({ App }, next) => {
        const getRouteApp = () => {
          if (isBrow) {
            return (props: any) => (
              <Router history={history}>
                {routesConfig ? (
                  renderRoutes(routesConfig, props)
                ) : (
                  <App {...props} />
                )}
              </Router>
            );
          }

          return (props: any) => {
            const runtimeContext = useContext(RuntimeReactContext);
            const basename = resolveBasename(historyOptions?.basename);
            const location = getLocation(runtimeContext?.ssrContext);
            const ctx = runtimeContext?.ssrContext?.redirection || {};

            return (
              <StaticRouter
                basename={basename}
                location={location}
                context={ctx}>
                {routesConfig ? (
                  renderRoutes(routesConfig, props)
                ) : (
                  <App {...props} />
                )}
              </StaticRouter>
            );
          };
        };

        return next({
          App: App ? hoistNonReactStatics(getRouteApp(), App) : getRouteApp(),
        });
      },
    }),
    { name: `@modern-js/plugin-router` },
  );
};
