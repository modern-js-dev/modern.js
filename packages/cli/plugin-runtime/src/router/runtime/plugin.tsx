import React, { useContext } from 'react';
import { createBrowserHistory, createHashHistory, History } from 'history';
import {
  RouteProps,
  unstable_HistoryRouter as HistoryRouter,
} from 'react-router-dom';
import { StaticRouter } from 'react-router-dom/server';
import { RuntimeReactContext } from '@modern-js/runtime-core';
import type { Plugin } from '@modern-js/runtime-core';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { isBrowser } from '../../common';
import { renderRoutes, getLocation, urlJoin } from './utils';
import { ServerRouterContext } from './context';

declare global {
  interface Window {
    _SERVER_DATA?: {
      router: {
        baseUrl: string;
        params: Record<string, string>;
      };
    };
  }
}

export type SingleRouteConfig = RouteProps & {
  redirect?: string;
  routes?: SingleRouteConfig[];
  key?: string;

  /**
   * layout component
   */
  layout?: React.ComponentType;

  /**
   * component would be rendered when route macthed.
   */
  component?: React.ComponentType;
};

export interface HistoryOptions {
  basename?: string;
  [key: string]: any;
}

export type HistoryConfig = {
  supportHtml5History?: boolean;
  historyOptions?: HistoryOptions;
};

export type RouterConfig = Partial<HistoryConfig> & {
  routesConfig?: {
    globalApp?: React.ComponentType<any>;
    routes?: SingleRouteConfig[];
  };
  history?: History;
  serverBase?: string[];
};

export const routerPlugin = ({
  serverBase = [],
  history: customHistory,
  supportHtml5History = true,
  routesConfig,
  // TODO: diff with history options 4 and 5
  historyOptions = {},
}: RouterConfig): Plugin => {
  const isBrow = isBrowser();

  const select = (pathname: string) =>
    serverBase.find(baseUrl => pathname.search(baseUrl) === 0) || '/';
  return {
    name: '@modern-js/plugin-router',
    setup: () => {
      return {
        hoc: ({ App }, next) => {
          const getRouteApp = () => {
            if (isBrow) {
              const baseUrl =
                window._SERVER_DATA?.router.baseUrl ||
                select(location.pathname);
              historyOptions.basename =
                baseUrl === '/'
                  ? urlJoin(baseUrl, historyOptions.basename as string)
                  : baseUrl;

              const history =
                customHistory ||
                (supportHtml5History
                  ? createBrowserHistory({ window })
                  : createHashHistory({ window }));

              // https://reactrouter.com/docs/en/v6/routers/history-router
              return (props: any) => (
                <HistoryRouter
                  basename={historyOptions.basename}
                  history={history}
                >
                  <App {...props}>
                    {routesConfig ? renderRoutes(routesConfig, props) : null}
                  </App>
                </HistoryRouter>
              );
            }
            return (props: any) => {
              const runtimeContext = useContext(RuntimeReactContext);
              const { ssrContext } = runtimeContext;
              const location = getLocation(ssrContext);
              const routerContext = ssrContext?.redirection || {};
              const request = ssrContext?.request;
              const baseUrl = request?.baseUrl as string;
              const basename =
                baseUrl === '/'
                  ? urlJoin(baseUrl, historyOptions.basename as string)
                  : baseUrl;
              return (
                // StaticRouter props context had been removed at v6
                // https://github.com/remix-run/react-router/issues/8406
                <ServerRouterContext.Provider value={routerContext}>
                  <StaticRouter
                    basename={basename === '/' ? '' : basename}
                    location={location}
                  >
                    <App {...props}>
                      {routesConfig ? renderRoutes(routesConfig, props) : null}
                    </App>
                  </StaticRouter>
                </ServerRouterContext.Provider>
              );
            };
          };
          let RouteApp = getRouteApp();

          if (App) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-expect-error hoist-non-react-statics types not adaptor for react 18
            RouteApp = hoistNonReactStatics(RouteApp, App);
          }

          if (routesConfig?.globalApp) {
            return next({
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-expect-error hoist-non-react-statics types not adaptor for react 18
              App: hoistNonReactStatics(RouteApp, routesConfig.globalApp),
            });
          }

          return next({
            App: RouteApp,
          });
        },
      };
    },
  };
};
