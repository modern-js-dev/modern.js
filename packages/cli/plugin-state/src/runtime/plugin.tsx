// eslint-disable-next-line filenames/match-exported
import { useContext } from 'react';
import { createPlugin, RuntimeReactContext } from '@modern-js/runtime-core';
import { createStore, Store } from '@modern-js-reduck/store';
import { Provider } from '@modern-js-reduck/react';
import hoistNonReactStatics from 'hoist-non-react-statics';

declare module '@modern-js/runtime-core' {
  interface RuntimeContext {
    store: Store;
  }

  interface TRuntimeContext {
    store: Store;
  }

  interface SSRData {
    storeState: any;
  }
}

type PluginProps = Parameters<typeof createStore>[0];

const state = (config: PluginProps) =>
  createPlugin(
    () => ({
      hoc({ App }, next) {
        const getStateApp = (props: any) => {
          // eslint-disable-next-line react-hooks/rules-of-hooks
          const context = useContext(RuntimeReactContext);

          return (
            <Provider store={context.store} config={config}>
              <App {...props} />
            </Provider>
          );
        };
        return next({
          App: hoistNonReactStatics(getStateApp, App),
        });
      },
      init({ context }, next) {
        const storeConfig = config || {};

        if (typeof window !== 'undefined') {
          storeConfig.initialState =
            storeConfig.initialState ||
            window?._SSR_DATA?.data?.storeState ||
            {};
        }

        context.store = createStore(storeConfig);

        next({ context });
      },
      pickContext({ context, pickedContext }, next) {
        return next({
          context,
          pickedContext: {
            ...pickedContext,
            store: context.store,
          },
        });
      },
    }),
    {
      name: '@modern-js/plugin-state',
    },
  );

export default state;

export * from '../plugins';
