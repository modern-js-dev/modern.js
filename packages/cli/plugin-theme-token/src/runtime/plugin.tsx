/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
// eslint-disable-next-line filenames/match-exported
import { createPlugin } from '@modern-js/runtime-core';
import React, { useContext } from 'react';

export const ThemeTokenContext = React.createContext({});

export const useThemeToken = (): ITokens =>
  useContext<ITokens>(ThemeTokenContext);

const themeToken = (
  options: {
    token?: Record<string, any>;
    useStyledComponentsThemeProvider?: boolean;
  } = {},
) =>
  createPlugin(
    () => ({
      hoc({ App }, next) {
        return next({
          App: (props: any) => {
            const { token = {}, useStyledComponentsThemeProvider = false } =
              options;
            if (useStyledComponentsThemeProvider) {
              const {
                ThemeProvider,
              } = require('@modern-js/runtime-core/styled');
              return (
                <ThemeProvider theme={token}>
                  <ThemeTokenContext.Provider value={token}>
                    <App {...props} />
                  </ThemeTokenContext.Provider>
                </ThemeProvider>
              );
            } else {
              return (
                <ThemeTokenContext.Provider value={token}>
                  <App {...props} />
                </ThemeTokenContext.Provider>
              );
            }
          },
        });
      },
    }),
    {
      name: '@modern-js/plugin-theme-token',
    },
  );

export default themeToken;

// export * from '../plugins';
/* eslint-enable @typescript-eslint/no-require-imports */
/* eslint-enable @typescript-eslint/no-var-requires */
