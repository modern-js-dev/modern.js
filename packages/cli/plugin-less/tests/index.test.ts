import plugin from '../src';
import plugin2 from '../src/cli';

describe('plugin-less', () => {
  it('default', () => {
    expect(plugin).toBeDefined();
    expect(plugin).toBe(plugin2);
  });
});
