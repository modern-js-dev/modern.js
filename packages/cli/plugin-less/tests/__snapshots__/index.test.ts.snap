// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin less test config 1`] = `
Array [
  Object {
    "tools": Object {
      "webpackChain": [Function],
    },
  },
]
`;

exports[`plugin less test schema 1`] = `
Array [
  Array [
    Object {
      "schema": Object {
        "typeof": Array [
          "object",
          "function",
        ],
      },
      "target": "tools.less",
    },
  ],
]
`;
